Fix bugs with aggregated discovery

Jefftree opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->
/kind bug

#### What this PR does / why we need it:

Fixes bug and test for new gated aggregated discovery feature which is disabled by default.

1.  When feature is enabled, aggregated api servers which only implement legacy discovery won't be included in the response document. This is because the code expects the wrong response status: NotFound was correct in the initial design. This was unfortunately not updated to become NotAvailable after the recent redesign during api review to use content-type negotiation on /apis.

2. Changed the conversion from APIResource to APIResourceDiscovery to use pointers instead of values. We were referencing array values as pointers which may change when the array get resized.

3. Sort versions based on VersionPriority. This needs to be updated from our basic reverse alpha sort.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

Jefftree commented:
/test all

Jefftree commented:
/retest

Jefftree commented:
/retest

Jefftree committed:
fix dropped subresources

some subresources were not properly included in the array due to pointers becoming stale over a resize

Jefftree committed:
add aggregated /apis handler for genericapiserver

so that aggregated-apiservers can also take advantage. discovered by e2e tests with feature enabled

alexzielenski committed:
fix aggregated discovery version sorting

add test for level based priorities

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/endpoints/installer.go:
```diff
@@ -126,12 +127,11 @@ func ConvertGroupVersionIntoToDiscovery(list []metav1.APIResource) ([]apidiscove
-			apiResourceList = append(apiResourceList, apidiscoveryv2beta1.APIResourceDiscovery{
+			parent = &apidiscoveryv2beta1.APIResourceDiscovery{
 				Resource: split[0],
 				Scope:    scope,
-			})
-			parentResources[split[0]] = &apiResourceList[len(apiResourceList)-1]
-			parent = &apiResourceList[len(apiResourceList)-1]
+			}
+			apiResourceList = append(apiResourceList, parent)
 			parentResources[split[0]] = parent
```
I dont think it was necessary to change the whole array to pointers. Rather than assigning parent to the address of an index you can assign it to the address of a variable similar to something like:

```go
parent = &apidiscoveryv2beta1.APIResourceDiscovery{...}
apiResourceList = append(apiResourceList, *parent)
```

Jefftree replied:
If we append to apiResourceList by value rather than by pointer, we cannot directly update the parent resource (in the subresources loop) without searching for the parent object in the array. That seems inefficient.

alexzielenski replied:
can we just store the index then?

Jefftree replied:
Yep, updated

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/handler_test.go:
```diff
@@ -402,6 +406,20 @@ func TestConcurrentRequests(t *testing.T) {
 
+// Shows that versions within groups are ordered in the dicvoery document
+// by their assigned version priority
+func TestVersionOrdering(t *testing.T) {
```
I think this is an incomplete version of the test `TestVersionSortingWithPriority`

alexzielenski commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go:
```diff
@@ -390,6 +393,8 @@ func (dm *discoveryManager) syncAPIService(apiServiceName string) error {
 	}
 
 	dm.mergedDiscoveryHandler.AddGroupVersion(gv.Group, entry)
+	dm.mergedDiscoveryHandler.SetGroupPriority(gv.Group, info.groupPriority)
```
Im curious what the behavior is of today's discovery when two APIServices have disagreeing `GroupPriorityMinimum`. Which wins?

alexzielenski replied:
Looks like it is the group with the higher minimum priority, as the name implies. This PR should account for that

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/handler.go:
```diff
@@ -164,13 +183,19 @@ func (rdm *resourceDiscoveryManager) addGroupVersionLocked(groupName string, val
 func (rdm *resourceDiscoveryManager) RemoveGroupVersion(apiGroup metav1.GroupVersion) {
 	rdm.lock.Lock()
 	defer rdm.lock.Unlock()
+	delete(rdm.versionPriorities, apiGroup)
```
can this be grouped with the `delete` lines?

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/handler.go:
```diff
@@ -95,6 +102,18 @@ func (rdm *resourceDiscoveryManager) SetGroupPriority(group string, priority int
+		rdm.versionPriorities[gv] = priority
+		rdm.cache.Store(nil)
+	} else {
+		klog.Warningf("DiscoveryManager: Attempted to set priority for group-version %s but does not exist", gv)
```
utilruntime.HandleError please

deads2k commented:
/approve

please squash to something tidier beforemerge
/assign @alexzielenski 

I'll leave lgtm with @alexzielenski unless something substantial changes.

alexzielenski committed:
add discovery integration tests

add crd integration tests
\

alexzielenski commented:
/milestone 1.26

bugfixes to alpha-gated feature + test additions

alexzielenski committed:
fix apiregistration priority

v1 discovery always has apiregistration first. this rigs the priority to be greater than maximum so it is guaranteed to be first in the list

alexzielenski committed:
fix aggregated discovery legacy fallback

due to redesign where we changed from new endpoint to /apis. The expected error was not also changed.

Caught by e2e tests when feature enabled. Should have been caught by unit test first but it was implemented without root /apis. Unit test also fixed

apelisse commented on test/integration/apiserver/discovery/discovery_test.go:
```diff
@@ -285,188 +222,437 @@ func TestAggregatedAPIServiceDiscovery(t *testing.T) {
-// -  Check that CRD is in discovery document
-func TestOverlappingCRDAndAPIService(t *testing.T) {
-	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.AggregatedDiscoveryEndpoint, true)()
+func runTestCases(t *testing.T, cases []testCase) {
```
Should this be a method of `testCase` that returns an error?

apelisse replied:
i.e.
```suggestion
func (tc testCase) Run() error {
```

apelisse replied:
That's nit by the way, the framework is great and totally reasonable for the current release stage.

deads2k commented:
/milestone 1.26

deads2k commented:
/milestone v1.26

alexzielenski commented:
/lgtm
